integration/lsp/viewengine_spec.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {MessageConnection} from 'vscode-jsonrpc';
+import * as lsp from 'vscode-languageserver-protocol';
+import {APP_COMPONENT, createConnection, FOO_TEMPLATE, initializeServer, openTextDocument} from './test_utils';
+
+describe('Angular Language Server', () => {
+  jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000; /* 10 seconds */
+
+  let client: MessageConnection;
+
+  beforeEach(() => {
+    client = createConnection({
+      ivy: false,
+    });
+    client.listen();
+  });
+
+  afterEach(() => {
+    client.dispose();
+  })
+
+  it('should handle initialization at startup', async () => {
+    const response = await initializeServer(client);
+    expect(response).toEqual({
+      capabilities: {
+        textDocumentSync: 2,
+        completionProvider: {
+          resolveProvider: false,
+          triggerCharacters: ['<', '.', '*', '[', '(', '$', '|'],
+        },
+        definitionProvider: true,
+        typeDefinitionProvider: true,
+        hoverProvider: true,
+        workspace: {
+          workspaceFolders: {
+            supported: true,
+          },
+        },
+      }
+    });
+  });
+
+  it('should handle completion', async () => {
+    await initializeServer(client);
+    openTextDocument(client, APP_COMPONENT);
+    client.sendNotification(lsp.DidChangeTextDocumentNotification.type, {
+      textDocument: {
+        uri: `file://${APP_COMPONENT}`,
+        version: 2,
+      },
+      contentChanges: [
+        {
+          range: {
+            start: {line: 4, character: 29},
+            end: {line: 4, character: 29},
+          },
+          text: '.',
+        },
+      ],
+    });
+    const response = await client.sendRequest(lsp.CompletionRequest.type, {
+      textDocument: {
+        uri: `file://${APP_COMPONENT}`,
+      },
+      position: {line: 4, character: 30},
+    });
+    expect(response).toContain({
```

Comment:
I think these expectations are fine for a first pass, but let's keep an eye on these for when there are more tests. These expectations aren't easy to understand and make it hard to derive meaning from them. Here are a couple things that I can think that would improve them with a `humanize` that does:
* Convert text ranges to actual text
* convert `kind` to a meaningful string
* I don't know what the meaning is for `label`, `sortText`, or `newText`. Maybe we don't need to assert all of those
*  The meaning of `detail` is related to where the text changed; we might want to improve the ways in which we construct the `sendNotification` object to also be more human readable so one can easily see "Oh, I'm adding a '.' char after XYZ text (I don't want to open the file and go to the line/char to figure that out), and I would expect the result to be X based on that."

Reply:
Yeah, we should definitely revisit how these integration tests are done.
Initially, what I had in mind was more like golden tests, where we just dump the entire response to JSON.
In fact, that's what I did for the integration test in `@angular/angular`: https://github.com/angular/angular/tree/master/integration/language_service_plugin/goldens

Although that test looks very similar to this one, they're different in one key way:
The one in `@angular/angular` is for tsserver **plugin** whereas this is for the LSP server.

We should probably unify if not consolidate these tests.